keystone/common/api_filtering.py

```diff
@@ -0,0 +1,210 @@
+                    msg = ('Role \'%s\' does not exist.'
+                           % kwargs.get('role_id'))
+                    LOG.error(_LE(msg))
+                    raise exception.NotFound(msg)
+            else:
+                refs = [i for i in res['roles']]
+                for i in refs:
+                    if (i['name']) in CONF.api_filtering.roles_to_filter:
+                        res['roles'].remove(i)
+        return res
+    return wrapper
+
+
+def filter_users(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+
+        project_details = {}
+        res = f(*args, **kwargs)
+
+        is_admin_flag = is_admin(args[0], args[1])
+
+        if not is_admin_flag:
+            projects = list_projects(args[0], args[1])
+
+            for project in projects['projects']:
+                if project['name'] in CONF.api_filtering.projects_to_filter:
+                    project_details[project['name']] = project['id']
+
+            if kwargs is not None and kwargs.get('user_id') is not None:
+                if res['user'].get('default_project_id') is not None:
+                    if (res['user']['default_project_id']
+                            in project_details.values()):
+                        msg = ('User \'%s\' does not exist.'
+                               % kwargs.get('user_id'))
+                        LOG.error(_LE(msg))
+                        raise exception.NotFound(msg)
+            else:
+                refs = [i for i in res['users']]
+
+                for i in refs:
+                    if (i.get('default_project_id') is not None and
+                            i['default_project_id']
+                            in project_details.values()):
+                        res['users'].remove(i)
+
+                    if (i.get('name')
+                            in CONF.api_filtering.explicit_users_to_filter):
+                        res['users'].remove(i)
+        return res
+    return wrapper
+
+
+def get_roles(self, context):
+    try:
+        user_token_ref = token_model.KeystoneToken(
+            token_id=context['token_id'],
+            token_data=self.token_provider_api.validate_token(
+                context['token_id']))
+    except exception.TokenNotFound as e:
+        LOG.error(_LE('Token not found'))
+        raise exception.Unauthorized(e)
+
+    creds = copy.deepcopy(user_token_ref.metadata)
+
+    try:
+        creds['user_id'] = user_token_ref.user_id
+    except exception.UnexpectedError:
+        LOG.error(_LE('Invalid user'))
+        raise exception.Unauthorized()
+
+    LOG.debug(_LE("user_token_ref project_scoped = %s "
+                  % user_token_ref.project_scoped))
+
+    if user_token_ref.project_scoped:
+        creds['tenant_id'] = user_token_ref.project_id
+    else:
+        LOG.error(_LE('Invalid tenant / project not scoped'))
+
+    creds['roles'] = user_token_ref.role_names
+
+    return creds['roles']
+
+
+def is_admin(self, context):
```

Comment:
This helper function should go down with list_projects. Also given that it's only intended to be used by other functions in the module it might be a good idea to pre-pend a `_` to the function name to indicate that it's "private".